### PR TITLE
restore: write index kv pairs and data kv pairs to separate engine

### PR DESCRIPTION
**Pull request by lonng**
### What problem does this PR solve? 

Write index kvs and data kvs to separate engine. See: #137 

### What is changed and how it works?

Use different engine file.

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

Related changes

### CODE REVIEW

**Comment by lonng** [comment]
@kennytm PTAL

**Comment by GregoryIan** [comment]
can we creare an issue for Proposal#4 ？

**Comment by lonng** [comment]
@GregoryIan PTAL

File: lightning/restore/restore.go [hunk 1]
```diff
@@ -339,19 +348,42 @@ func (rc *RestoreController) listenCheckpointUpdates(wg *sync.WaitGroup) {
...
 		// 	panic("forcing failure due to FailIfStatusBecomes")
 		// }
-		// continue
+		// goto gofailLabel2
+	gofailLabel2:
+		if false {
+			goto gofailLabel2
+		}
+
+		// gofail: var FailIfIndexEngineImported int
```
Comment (GregoryIan) [inline lightning/restore/restore.go]: it seems failpoint `FailIfStatusBecomes` is able to take place of `FailIfIndexEngineImported `

File: lightning/restore/restore.go [hunk 2]
```diff
@@ -529,38 +597,50 @@ func (t *TableRestore) restoreTable(
...
 		common.AppLogger.Infof("[%s] import whole table takes %v", t.tableName, time.Since(timer))
 		err := engineErr.Get()
-		rc.saveStatusCheckpoint(t.tableName, -1, err, CheckpointStatusImported)
 		if err != nil {
 			return errors.Trace(err)
 		}
+
+		// If index engine file has been closed but not imported only if context cancel occurred
+		// when `importKV()` execution, so `UnsafeCloseEngine` and continue import it.
+		if indexEngineCp.Status == CheckpointStatusClosed {
```
Comment (GregoryIan) [inline lightning/restore/restore.go]: what if lightning execute L793 (L794 doesn't take effect), and then lightning was killed directly?

File: lightning/config/config.go [hunk 3]
```diff
@@ -229,11 +230,17 @@ func (cfg *Config) Load() error {
...
-		*cfg.PostRestore.Level1Compact = true
+		*cfg.PostRestore.Level1Compact = false
 	}
 
+	if cfg.App.TableConcurrency < 2 {
+		cfg.App.TableConcurrency = 2
+	}
+	if cfg.App.IndexConcurrency < 2 {
+		cfg.App.IndexConcurrency = 2
+	}
```
Comment (kennytm) [inline lightning/config/config.go]: I don't think you need to mandate the minimal concurrency at all

File: lightning/restore/file_checkpoints.proto [hunk 4]
```diff
@@ -28,12 +28,14 @@ message TableCheckpointModel {
...
     int64 alloc_base = 4;
     repeated EngineCheckpointModel engines = 6;
+    repeated uint32 index_engine_statuses = 7;
 }
 
 message EngineCheckpointModel {
     uint32 status = 1;
     // key is "$path:$offset"
     map<string, ChunkCheckpointModel> chunks = 2;
+    int32 engine_id = 3;
```
Comment (kennytm) [inline lightning/restore/file_checkpoints.proto]: You shouldn't need engine_id at all! In RFC 4 the status should be written into the `index_engine_statuses` field!

File: lightning/restore/restore.go [hunk 5]
```diff
@@ -339,19 +348,42 @@ func (rc *RestoreController) listenCheckpointUpdates(wg *sync.WaitGroup) {
...
 		// if _, ok := scp.merger.(*ChunkCheckpointMerger); ok {
 		// 	wg.Wait()
 		// 	panic("forcing failure due to FailIfImportedChunk")
 		// }
-		// continue
+		// goto gofailLabel1
+	gofailLabel1:
+		if false {
+			goto gofailLabel1
+		}
```
Comment (kennytm) [inline lightning/restore/restore.go]: You don't need these `if false` stuff, see the function `OpenEngine` in `kv/importer.go`.

File: lightning/config/config.go (hunk 3 above)
Comment (lonng) [reply lightning/config/config.go]: adressed

File: lightning/restore/checkpoints.go [hunk 6]
```diff
@@ -108,8 +109,9 @@ type ChunkCheckpoint struct {
...
 
 type EngineCheckpoint struct {
-	Status CheckpointStatus
-	Chunks []*ChunkCheckpoint // a sorted array
+	EngineID int
+	Status   CheckpointStatus
+	Chunks   []*ChunkCheckpoint // a sorted array
 }
 
 type TableCheckpoint struct {
```
[the remark this answers is not in the document]
Comment (lonng) [reply lightning/restore/checkpoints.go]: Actually, we should treat it as a normal Engine.

File: lightning/restore/checkpoints.go [hunk 7]
```diff
@@ -192,7 +194,7 @@ func (merger *StatusCheckpointMerger) SetInvalid() {
 }
 
 func (merger *StatusCheckpointMerger) MergeInto(cpd *TableCheckpointDiff) {
-	if merger.EngineID == -1 || merger.Status <= CheckpointStatusMaxInvalid {
+	if merger.EngineID == invalidEngineID || merger.Status <= CheckpointStatusMaxInvalid {
```
[the remark this answers is not in the document]
Comment (lonng) [reply lightning/restore/checkpoints.go]: addressed

File: lightning/restore/restore.go (hunk 1 above)
Comment (lonng) [reply lightning/restore/restore.go]: We need an additional failpoint to cover FailIfIndexEngineImported also. See `tests/checkpoint`

File: lightning/restore/restore.go (hunk 5 above)
Comment (lonng) [reply lightning/restore/restore.go]: addressed

File: lightning/restore/restore.go (hunk 2 above)
Comment (lonng) [reply lightning/restore/restore.go]: addressed

File: lightning/restore/restore.go [hunk 8]
```diff
@@ -54,6 +56,11 @@ const (
 	defaultGCLifeTime = 100 * time.Hour
 )
 
+const (
+	indexEngineID   = -1
+	invalidEngineID = math.MinInt64
```
[the remark this answers is not in the document]
Comment (lonng) [reply lightning/restore/restore.go]: addressed

File: lightning/restore/restore.go [hunk 9]
```diff
@@ -501,14 +533,45 @@ func (t *TableRestore) restoreTable(
...
+			defer rc.indexWorkers.Recycle(indexWorker)
+			var err error
+			indexEngine, err = rc.importer.OpenEngine(ctx, t.tableName, indexEngineID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
 
+		// The table checkpoint status less than `CheckpointStatusIndexImported` implies
+		// that index engine checkpoint status less than `CheckpointStatusImported`.
```
[the remark this answers is not in the document]
Comment (lonng) [reply lightning/restore/restore.go]: addressed

**Comment by GregoryIan** [comment]
LGTM

File: lightning/config/config.go [hunk 10]
```diff
@@ -229,11 +231,10 @@ func (cfg *Config) Load() error {
 		}
 	}
 
-	// If the level 1 compact configuration not found, default to true
+	// If the level 1 compact configuration not found, default to false
```
Comment (kennytm) [inline lightning/config/config.go]: If we're changing the default to `false` it doesn't need the pointer indirection anymore (but why changing it to `false` in this PR?)

File: lightning/restore/file_checkpoints.proto (hunk 4 above)
Comment (kennytm) [reply lightning/restore/file_checkpoints.proto]: Let's discuss about this tomorrow morning

**Pull request merged**